subsys/bluetooth/controller/ll_sw/ull_sync.c

```diff
@@ -0,0 +1,630 @@
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_PHY_CODED)) {
+		scan_coded = ull_scan_set_get(SCAN_HANDLE_PHY_CODED);
+		if (!scan_coded || scan_coded->per_scan.sync) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
+	}
+
+	link_sync_estab = ll_rx_link_alloc();
+	if (!link_sync_estab) {
+		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+	}
+
+	link_sync_lost = ll_rx_link_alloc();
+	if (!link_sync_lost) {
+		ll_rx_link_release(link_sync_estab);
+
+		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+	}
+
+	node_rx = ll_rx_alloc();
+	if (!node_rx) {
+		ll_rx_link_release(link_sync_lost);
+		ll_rx_link_release(link_sync_estab);
+
+		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+	}
+
+	sync = sync_acquire();
+	if (!sync) {
+		ll_rx_release(node_rx);
+		ll_rx_link_release(link_sync_lost);
+		ll_rx_link_release(link_sync_estab);
+
+		return BT_HCI_ERR_MEM_CAPACITY_EXCEEDED;
+	}
+
+	scan->per_scan.filter_policy = options & BIT(0);
+	if (!scan->per_scan.filter_policy) {
+		scan->per_scan.sid = sid;
+		scan->per_scan.adv_addr_type = adv_addr_type;
+		memcpy(scan->per_scan.adv_addr, adv_addr, BDADDR_SIZE);
+	}
+
+	sync->skip = skip;
+	sync->timeout = sync_timeout;
+
+	/* TODO: Support for CTE type */
+
+	/* Initialize sync context */
+	sync->timeout_reload = 0U;
+	sync->timeout_expire = 0U;
+	lll_sync = &sync->lll;
+	lll_sync->skip_prepare = 0U;
+	lll_sync->skip_event = 0U;
+	lll_sync->data_chan_id = 0U;
+	lll_sync->window_widening_prepare_us = 0U;
+	lll_sync->window_widening_event_us = 0U;
+
+	/* Reporting initially enabled/disabled */
+	lll_sync->is_enabled = options & BIT(1);
+
+	/* Initialise state */
+	scan->per_scan.state = LL_SYNC_STATE_IDLE;
+
+	/* established and sync_lost node_rx */
+	node_rx->link = link_sync_estab;
+	scan->per_scan.node_rx_estab = node_rx;
+	sync->node_rx_lost.link = link_sync_lost;
+
+	/* Initialise ULL and LLL headers */
+	ull_hdr_init(&sync->ull);
+	lll_hdr_init(lll_sync, sync);
+
+	/* Enable scanner to create sync */
+	scan->per_scan.sync = sync;
+	if (IS_ENABLED(CONFIG_BT_CTLR_PHY_CODED)) {
+		scan_coded->per_scan.sync = sync;
+	}
+
+	return 0;
+}
+
+uint8_t ll_sync_create_cancel(void **rx)
+{
+	struct ll_scan_set *scan_coded;
+	memq_link_t *link_sync_estab;
+	memq_link_t *link_sync_lost;
+	struct node_rx_pdu *node_rx;
+	struct ll_scan_set *scan;
+	struct ll_sync_set *sync;
+	struct node_rx_sync *se;
+
+	scan = ull_scan_set_get(SCAN_HANDLE_1M);
+	if (!scan || !scan->per_scan.sync) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
+
+	if (IS_ENABLED(CONFIG_BT_CTLR_PHY_CODED)) {
+		scan_coded = ull_scan_set_get(SCAN_HANDLE_PHY_CODED);
+		if (!scan_coded || !scan_coded->per_scan.sync) {
+			return BT_HCI_ERR_CMD_DISALLOWED;
+		}
+	}
+
+	sync = scan->per_scan.sync;
+	scan->per_scan.sync = NULL;
+	if (IS_ENABLED(CONFIG_BT_CTLR_PHY_CODED)) {
+		scan_coded->per_scan.sync = NULL;
+	}
+
+	/* Check for race condition where in sync is established when sync
+	 * context was set to NULL.
+	 */
+	if (sync->timeout_reload) {
+		return BT_HCI_ERR_CMD_DISALLOWED;
+	}
```

Comment:
`sync->timeout_reload` is used to detect if a sync was established. This member is initialized to 0, but gets a valid value based on the periodic interval on sync establishment.
Thinking aloud, `scan->per_scan.sync = NULL;` should prevent any establishment race. Either the LLL ISR runs before this or after this. If run before, then sync is established and sync cancel command shall fail. Or if LLL ISR runs after, then it will not establish sync.

Ok. There is a missing NULL check.

```suggestion
	/* Check for race condition where in sync is established when sync
	 * context was set to NULL.
	 */
	if (!sync || sync->timeout_reload) {
		return BT_HCI_ERR_CMD_DISALLOWED;
	}
```

I got some new grey hair replying to justify the race condition handling ;-)

Reply:
I'm not sure I understand the issue nor the solution properly yet :) 

If there is indeed another thread that may change the `sync->timeout_reload`, adding this `if` does not really fix anything as far as I can tell, since there's no way of knowing if this `if` statement will be run before or after the other thread changes that value. 

If you have a single variable that is modified or read by multiple threads, wouldn't the proper way to handle that to be adding a mutex or using an atomic block?

Reply:
Two variables are used, first `scan->per_scan.sync` is set to `NULL`, a race here is acceptable as the desired final value is `NULL`. Secondly, `sync->timeout_reload` is checked in thread mode. If ISR runs before this check, and before `scan->per_scan.sync = NULL;`, then sync will have been established and `sync->timeout_reload` will not be `0`. If the ISR runs after the check, then sync will not be established in the LLL ISR due to the `scan->per_scan.sync = NULL;`.

Reply:
I see. 

So if ISR runs when `scan->per_scan.sync` isn't NULL, it'll set `sync->timeout_reload` which essentially means that the sync is established. 

So there's two common variables here, but the latter relies on the former (`scan->per_scan.sync`). 

I wonder if something like `atomic_ptr_cas` would be useful here. If both this thread and ISR use that when checking `scan->per_scan.sync`, can't we avoid the race condition?

Reply:
`atomic_ptr_cas`  uses irq_lock, that is something avoided in the controller design.